### PR TITLE
Keccak256 section in the wiki

### PR DESCRIPTION
**Pull request by DeluxeRaph**
Adding keccak section focusing on the following aspects:
- A brief overview of Keccak's development, highlighting its origins.
- The introduction of the altered Keccak256 variant specifically created for the Ethereum Virtual Machine (EVM).
- An exploration of the various use cases of Keccak256 within the architecture of Ethereum, detailing its significance and applications.

### CODE REVIEW

File: docs/wiki/Cryptography/keccak256.md [hunk 1]
```diff
@@ -0,0 +1,23 @@
+# Keccak256
+
+Keccak256 is a cryptographic hash function prominently used in the Ethereum blockchain. 
+
+## Brief History
+
+Keccak256 originated as a submission to the [NIST cryptographic hash algorithm competition](https://csrc.nist.gov/projects/hash-functions/nist-hash-function-competition). This competition aimed to identify a new secure hash algorithm to replace SHA-1. The KECCAK algorithm was designed by a team comprising Guido Bertoni, Joan Daemen, Michaël Peeters, and Gilles Van Assche. Their design was notable for its ability to absorb a phrase of arbitrary length and produce a fixed-size output.
```
Comment (taxmeifyoucan) [inline docs/wiki/Cryptography/keccak256.md:7]: > Their design was notable for its ability to absorb a phrase of arbitrary length and produce a fixed-size output.

This is property of most hash function and not really specific to keccak. Maybe clarify what's the difference between their original winning submission used in eth and final sha3 version.

File: docs/wiki/Cryptography/keccak256.md [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+
+- **Block Header Fields**: Various fields in the block header, such as `parentHash` and `stateRoot`, use the Keccak 256-bit hash. This includes hashing the entire header of the parent block, the root node of the state trie, and the root nodes of the trie structures for transactions and receipts.
+- **Merkle Patricia Tree**: Ethereum employs a Merkle Patricia Tree to encode its state, where each node in the tree is identified through the Keccak 256-bit hash of its content. This structure underpins the stateRoot field in the block header.
+- **Storage Contents Encoding**: The hash is used to encode the storage contents of accounts, mapping the Keccak 256-bit hash of integer keys to the RLP-encoded integer values.
+
+In all these instances, Keccak256's role is critical for ensuring data integrity, facilitating efficient data retrieval, and supporting the blockchain's underlying security mechanisms.
+
+## References
+- [NIST SHA-3 Competition](https://csrc.nist.gov/projects/hash-functions/nist-hash-function-competition)
+- [Ethereum Yellow Paper](https://ethereum.github.io/yellowpaper/paper.pdf)
```
Comment (taxmeifyoucan) [inline docs/wiki/Cryptography/keccak256.md:23]: I think a better resource would be the contest submission https://keccak.team/files/Keccak-submission-3.pdf

File: docs/wiki/Cryptography/keccak256.md [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+## EVM Implementation
+While SHA-3 is the official version of the algorithm, Ethereum utilizes a modified variant known as Keccak256. This version is tailored for a maximum input of 256 bits. The [Ethereum Virtual Machine (EVM)](https://ethereum.org/en/developers/docs/evm/), a Turing-complete virtual machine, employs Keccak256 for data hashing. 
+
+## Usage in Block Creation and Root Data Structure
+Keccak256 plays a vital role in Ethereum's block creation and in its root data structures. According to the [Ethereum Yellow Paper](https://ethereum.github.io/yellowpaper/paper.pdf), the function is used in several key ways:
+
+- **Block Header Fields**: Various fields in the block header, such as `parentHash` and `stateRoot`, use the Keccak 256-bit hash. This includes hashing the entire header of the parent block, the root node of the state trie, and the root nodes of the trie structures for transactions and receipts.
```
Comment (taxmeifyoucan) [inline docs/wiki/Cryptography/keccak256.md:15]: Check with @varun-doshi about his work on #28 which might be linked here. These docs should refer to each other but not repeat what was already documented.

File: docs/wiki/Cryptography/keccak256.md (hunk 2 above)
Comment (taxmeifyoucan) [reply docs/wiki/Cryptography/keccak256.md:23]: Here is also a good explanation you can use or link https://ethereum.stackexchange.com/questions/550/which-cryptographic-hash-function-does-ethereum-use

File: docs/wiki/Cryptography/keccak256.md [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+- **Merkle Patricia Tree**: Ethereum employs a Merkle Patricia Tree to encode its state, where each node in the tree is identified through the Keccak 256-bit hash of its content. This structure underpins the stateRoot field in the block header.
+- **Storage Contents Encoding**: The hash is used to encode the storage contents of accounts, mapping the Keccak 256-bit hash of integer keys to the RLP-encoded integer values.
+
```
Comment (taxmeifyoucan) [inline docs/wiki/Cryptography/keccak256.md:18]: You didn't mention it in EVM implementation before and it's necessary to also mention keccak EVM opcode

File: docs/wiki/Cryptography/keccak256.md (hunk 1 above)
Comment (taxmeifyoucan) [reply docs/wiki/Cryptography/keccak256.md:7]: You can also elaborate or add a section on keccak design, mention it's using a sponge function

**Review by taxmeifyoucan — COMMENTED** [review]
Thanks a lot for the submission! This is a great start, I added few comments

File: docs/wiki/Cryptography/keccak256.md (hunk 1 above)
Comment (raxhvl) [reply docs/wiki/Cryptography/keccak256.md:7]: Maybe also point out that even though the opcode is called `SHA3`, it computes Keccak-256 and the NOT the finalized standard (FIP-202 SHA-3). Related discussion: https://github.com/ethereum/EIPs/issues/59.

An empty vector is used to differentiate between them:
```
Keccak256("") =
  c5d2460186f7233c927e7db2dcc703c0e500b653ca82273b7bfad8045d85a470
```
```
SHA3("") =
  a7ffc6f8bf1ed76651c14756a061d662f580ff4de43b49fa82d80a4b80f8434a
```

File: docs/wiki/Cryptography/keccak256.md (hunk 3 above)
Comment (DeluxeRaph) [reply docs/wiki/Cryptography/keccak256.md:34]: Can I update it to @varun-doshi section after his section is updated?

Comment (taxmeifyoucan) [reply docs/wiki/Cryptography/keccak256.md:34]: sure, we can merge it and update later on, I don't want to block you with it

**Comment by DeluxeRaph** [comment]
I update the keccak design, evm opcode implementation and changed references

**Comment by taxmeifyoucan** [comment]
Thanks for the updates! All looks good to me

**Pull request merged**